### PR TITLE
Identify system threads using a Thread subclass

### PR DESCRIPTION
**Pull request by carlosdelest**
Part of the work in https://github.com/elastic/elasticsearch/pull/113194

Identifies system threads by using a dedicated `Thread` subclass, `EsThread`. 

`EsExecutor` includes a `isSystem` attribute that gets populated on creation according to the executor name. This attribute is used to flag the threads the executor creates.

A `isSystem` attribute in the `EsThread` class can be checked for understanding if a thread is a system thread.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java [hunk 1]
```diff
@@ -325,40 +334,59 @@ public static String executorName(Thread thread) {
...
         final AtomicInteger threadNumber = new AtomicInteger(1);
         final String namePrefix;
+        final boolean isSystem;
 
-        EsThreadFactory(String namePrefix) {
+        EsThreadFactory(String namePrefix, String executorName) {
             this.namePrefix = namePrefix;
             SecurityManager s = System.getSecurityManager();
             group = (s != null) ? s.getThreadGroup() : Thread.currentThread().getThreadGroup();
+            isSystem = executorName != null && SYSTEM_THREAD_PREFIXES.contains(executorName);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:366]: I know I did this too in my prototype, but you have to admit it's a little lame? Think we could maybe pass the `isSystem` property through the steps of building the executor via a boolean flag?

**Review by original-brownbear — COMMENTED** [review]
I like this. I think core-infra should also be ok with it. It would be good however if we could pass the `isSystem` value via a boolean. Not just for aesthetic reasons, but as mentioned I think there's other things we can do with subclassing `Thread` here and it would be nice to start in a clean way :)

**Comment by carlosdelest** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java (hunk 1 above)
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java]: Makes sense - removed dependency on naming in 71a0d285f82aaa40698197df3fd90eaafce49480

**Review by rjernst — APPROVED** [review]
Looks ok to me

**Comment by carlosdelest** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/115618)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**